lib/TokenManager.js

```diff
@@ -122,17 +143,38 @@ function refresh(sdk, tokenMgmtRef, storage, key) {
   // Remove existing autoRefresh timeout for this key
   clearRefreshTimeout(tokenMgmtRef, key);
 
+  var deferred = Q.defer();
```

Comment:
The concern here: if we return the same promise, then multiple callers will be chaining off the same promise, which might result in errors from other callers being bubbled upwards in unexpected ways?

Reply:
Is there a specific example you're concerned about?

In general, promise resolution is decoupled from handlers, and handlers are independent from each other. When a promise is resolved, it calls any handlers that have been added to that promise (which is what this code is also trying to do).

Callers are chaining off the same promise, but they are not chaining off each other (which would result in errors bubbling up, but that's expected) i.e. it's not a two way error-propagation street.

```javascript
const somePromise = new Promise(resolve => resolve('someVal'));

somePromise
.then(() => {
  console.log('Chain 1: throw an error');
  throw new Error('Error!');
})
.then(() => {
  console.log('Chain 1: will never get here');
});

somePromise
.then(() => {
  console.log('Chain 2: different chain, will log');
});
```

Reply:
As mentioned offline: I had some misunderstanding of promises, so I see why it isn't necessary to maintain a list of promises to resolve with the same value.  But I do think it's important that we cache a singular promise while we're refreshing the token, so that we don't send out multiple refresh requests to the API (and get multiple tokens minted) if the consumer of this library has inadvertently called this method many times (as is likely if multiple UI components have a dependency on getting a token in order to do work)